### ISSUE TITLE
[Help] Facebook response jumping out of conversation scope

### ISSUE THREAD

**Issue by kniteshift** [howdyai/botkit#1173]
Botkit version: 0.5.8
Platform: Facebook

```js 
controller.hears([/^SelectItem\/*/i], 'message_received', function(bot, message) {
// this is the particular case that's running into an issue
if (item.options.length < 4) {
        item.options.forEach((option, index) => {
          if(option.price > 0) {
            buttons.push({
              "type": "postback",
              "title": option.price>0 ?`${option.name} $${option.price.toFixed(2)}`: '',
              "payload": index
            })
          }
        })
        msg = {
          attachment: {
            'type': 'template',
            'payload': {
              'template_type': 'generic',
              'elements': [
                {
                  'title': `Now, select your ${item.name}`,
                  "buttons": buttons
                }
              ]
            }
          }
        }
      }
})
```
Now within the same ``controller.hears`` scope I have the following: 
```js 
bot.startConversation(message, function(err, convo) {
        convo.addQuestion(msg, function(response, convo) {
          console.log('response', response)
          if (response) {
            let newItem = Object.assign({}, item);
            // changing this from reponse.payload to response.text as it does not seem like response object yields payload
            newItem.options = [item.options[response.text]];

         /* ...  bot should reply with something here */
        }
        else {
          orderHelper.addItemToOrder(bot, message, newItem);
        }
            convo.next();
          }
        })
      });
```

``response.text`` should be the index posted back from the webhook. However, it sometimes gets captured by the catchall below. 
```js
controller.hears('.*', 'message_received', function (bot, message) {
})
```

This behavior occurs some of the time, not all of the time, where it jumps out of the conversations scope.

**Comment by jonchurch** [comment]
Turn on debug mode by passing `debug: true` into your controller options, if you haven't already. 

That will tell you when a conversation begins and ends in your logs. It sounds like your conversation is exiting early somewhere, this only occurs after the hears trigger runs and the bot sends the user the buttons, right?

Try using convo.ask instead of convo.addQuestion, ask should run immediately where addQuestion is for composing a conversation and then starting it later. With bot.startConversation you're beginning the convo as soon as that runs, so maybe the question isn't making it onto the queue before the convo ends.

I don't understand what is happening with the `if (response) {} else {}` block, response should always be present when that handler runs.

**Comment by kniteshift** [comment]
@jonchurch it looks like this MAY be fixed. I took `require_delivery: true` off of the configuration, and it seems like the conversation is persisting. I will watch this for a few days but if all is good, then I am going to close this issue.

**Comment by CitiVince** [comment]
Version 0.6.7

Came here post a similar issue, jumping from one `controller.hears` scope to a different one on quick replies. The behavior is not consistent, i.e. from the same point in the original conversation, using the same quick reply it jumps to different messages in different `convo.hears` scopes or stays in scope and follows the script correctly.

**Comment by CitiVince** [comment]
Hijacking this issue to not open a new one. @jonchurch is it possible that botkit will jump out of `hear` scopes when using arrow functions instead of regular function declarations? 

What I had (probably source of error)
````javascript
    controller.hears(['TESTIT'], 'message_received', (bot, message) => {
          bot.createConversation(/*...*/)
    })
````
Now switched to what is described in the documentation:
````javascript
    controller.hears(['TESTIT'], 'message_received', function(bot, message) {
          bot.createConversation(/*...*/)
    })
````
And it _seems_ to solve the problem of jumping accross `threads` and `hear` scopes. Can it be that this difference has such a dramatic impact?

Feedback would be very much appreciated!

**Comment by kniteshift** [comment]
@CitiVince I don't think this would be the case, because there is no usage of lexical context on keyword `this` for bot conversation. 

On my end, the bot seems to be staying within conversation now that I removed `required_delivery: true` in the botkit config. I did notice that a few cases came from when we were trying to access the storage module as well, where the conversation seems to jump before the data fetch from db resolved. I fixed this by promisfying using Node 8.0's promisfy method in `utils`, and will probably make a switch to bluebird this weekend.

I wonder though, for our cases, if we can use something like `async` functions to `await` each thread of the conversation.

**Comment by CitiVince** [comment]
@kniteshift cheers for your feedback!

I also noticed that `this` isn't used in `convo`. And as it turns out, the bot keeps jumping out of scope. However, we noticed a pattern probably relating to the code below. One obvious mistake was to not use `convo.stop()` after menu postback was heard. 

Still, we are not sure whether the approach below is the best way to properly exit a `convo` based on user inputs and move on with a clean slate. Any feedback on this would be very much appreciated!

````javascript
controller.middleware.capture.use((bot,message,convo, next) => {

    let menu_postbacks = [/* array of menu postbacks */]
  
    if (message.text === 'Stop'){
        // catch stop keyword
        convo.stop()
        bot.reply(convo.source_message, 'Ok')        
    } else if (message.postback && menu_postbacks.includes(message.postback.payload)){
        // catch menu postbacks mid convo
        convo.stop()  //<-- This was obviously missing
        controller.trigger('facebook_postback', [bot, message])
    } else {
        next()
    }
})
````

**Comment by kniteshift** [comment]
@jonchurch it seems that issue still persists. If I click the button multiple times after error, it woudl eventually resolve. It looks like my catch all gets in the way of the conversation.

**Comment by anglol** [comment]
Something new about this point ?

Got the same issue :(
